src/main/java/updater/JarUpdater.java

```diff
@@ -0,0 +1,275 @@
+    private final ExecutorService pool = Executors.newSingleThreadExecutor();
+
+    private static final int MAX_RETRY = 10;
+    private static final int WAIT_TIME = 2000;
+    private static final String BACKUP_FILENAME_SUFFIX = "_backup";
+    private static final String ERROR_ON_UPDATING =
+            "Update cannot be applied. Update will be aborted. Please close this window.";
+
+    private Label updatingLabel;
+
+    @SuppressWarnings("PMD")
+    // PMD - "Consider using varargs for methods or constructors which take an array the last parameter."
+    public static void main(String[] args) {
+        Application.launch(args);
+    }
+
+    @Override
+    public void start(Stage stage) {
+        initLogger();
+        showWaitingWindow(stage);
+        pool.execute(() -> run());
+    }
+
+    private void showWaitingWindow(Stage stage) {
+        stage.setTitle("Applying Updates");
+        VBox windowMainLayout = new VBox();
+        Group root = new Group();
+        Scene scene = new Scene(root);
+        stage.setScene(scene);
+        scene.setRoot(windowMainLayout);
+
+        updatingLabel = new Label();
+        updatingLabel.setText("Please wait. HubTurbo is being updated...");
+        updatingLabel.setPadding(new Insets(50));
+
+        windowMainLayout.getChildren().addAll(updatingLabel);
+
+        stage.show();
+    }
+
+    private void run() {
+        HashMap<String, String> commandLineArgs = new HashMap<>(getParameters().getNamed());
+
+        String sourceJarPath = commandLineArgs.get("source");
+        String targetJarPath = commandLineArgs.get("target");
+        String executeJarOption = commandLineArgs.get("execute-jar");
+
+        if (sourceJarPath == null || targetJarPath == null) {
+            log("Please specify source and target files.");
+            quit();
+        } else {
+            log("source: " + sourceJarPath);
+            log("target: " + targetJarPath);
+        }
+
+        File sourceJarFile = new File(sourceJarPath);
+        File targetJarFile = new File(targetJarPath);
+
+        if (!sourceJarFile.exists()) {
+            log("Such source file does not exist, aborting.");
+            quit();
+        }
+
+        if (!prepareTargetJarFile(targetJarFile)) {
+            showCriticalErrorOnUpdating();
+            return;
+        }
+
+        log("Moving source to target");
+        if (!moveFile(sourceJarFile.toPath(), targetJarFile.toPath())) {
+            log("Failed to move update.");
+            restoreBackup(targetJarFile);
+            quit();
+        }
+
+        // Optionally start target jar file
+        if (executeJarOption != null && executeJarOption.equalsIgnoreCase("y")) {
+            if (executeJar(targetJarPath, "")) {
+                removeBackup(targetJarFile);
+            } else { // if target can't be started, rollback backup
+                restoreBackup(targetJarFile);
+                executeJar(targetJarPath, "");
```

Comment:
Is it okay to ignore the return value of this? Maybe instead of having error-handling everywhere, we can use an exception and handle it higher up? I know this won't work for everything, but if all these stages should return/quit/abort on failure, an exception to stop execution there might work better.